crypto/fipsmodule/sha/internal.h

```diff
@@ -405,40 +416,128 @@ OPENSSL_EXPORT uint8_t *SHAKE128(const uint8_t *data, const size_t in_len,
+ *
+ * SHAKE functions never fail when the later call-discipline is adhered to:
+ * (a) the context execution flow is followed (b) the pointers passed to the
+ * functions are valid (c) any additional per-function parameter value conditions,
+ * detailed above each SHAKE_ function signature, is satisfied.
+ */
+
+// SHAKE_Init initialises |ctx| fields through |FIPS202_Init| and
+// returns 1 on success and 0 on failure. When call-discipline is
+// maintained and |block_size| value corresponds to a SHAKE block size length
+// in bytes, this function never fails.
 int SHAKE_Init(KECCAK1600_CTX *ctx, size_t block_size);
 
-// SHAKE_Absorb checks |ctx| pointer and |len| values. It updates and absorbs 
-// input blocks via |FIPS202_Update|.
+// SHAKE_Absorb checks |ctx| pointer and |len| values. It updates and absorbs
+// input blocks via |FIPS202_Update|. When call-discipline is
+// maintained and |len| value corresponds to the input message length
+// (including zero), this function never fails.
 int SHAKE_Absorb(KECCAK1600_CTX *ctx, const void *data,
                                size_t len);
 
-// SHAKE_Squeeze pads the last data block and absorbs it through 
-// |FIPS202_Finalize| on first call. It writes |len| bytes of incremental 
-// XOF output to |md| and returns 1 on success and 0 on failure. It can be 
-// called multiple times.
+// SHAKE_Squeeze pads the last data block and absorbs it through
+// |FIPS202_Finalize| on first call. It writes |len| bytes of incremental
+// XOF output to |md| and returns 1 on success and 0 on failure. It can be
+// called multiple times. When call-discipline is maintained, this function
+// never fails.
 int SHAKE_Squeeze(uint8_t *md, KECCAK1600_CTX *ctx, size_t len);
 
 // SHAKE_Final writes |len| bytes of finalized extendible output to |md|, returns 1 on
 // success and 0 on failure. It should be called once to finalize absorb and
 // squeeze phases. Incremental XOF output should be generated via |SHAKE_Squeeze|.
+// When call-discipline is maintained, this function never fails.
 int SHAKE_Final(uint8_t *md, KECCAK1600_CTX *ctx, size_t len);
 
+/*
+ * SHAKE128_x4_ batched APIs implement x4 SHAKE functionalities on top of FIPS202 API layer
+ *
+ * SHAKE128_x4_ context must go through the flow: (a) Init_x4, (b) Absorb_once_x4 [one time;
+ * maximum input length of |SHAKE128_BLOCKSIZE - 1|] (c) Squeezeblocks_x4 [multiple times]
+ *
+ * SHAKE128_x4_ functions never fail when the later call-discipline is adhered to:
+ * (a) the context execution flow is followed (b) the pointers passed to the
+ * functions are valid (c) any additional per-function parameter value conditions,
+ * detailed above each SHAKE128_x4_ function signature, is satisfied.
+ */
+
+// SHAKE128_Init_x4 is a batched API that operates on four independent
+// Keccak bitstates. It initialises all four |ctx| fields through four
+// consecutive calls to |SHAKE_Init| and returns 1 on success and 0 on failure.
```

Comment:
Can we remove the mention of the implementation details here and elsewhere? Once we start implementing fast batched x4, we won't merely fall back to x1 anymore.

Reply:
I tried to follow the general pattern for function description in the file/library. I was thinking once we start implementing it, we could update these comments.
I just fixed it. Thank you.